File: c/eu020.c
// project euler (projecteuler.net) problem 20
// solution by Kevin Retzke (retzkek@gmail.com)
// November 2012

#include <stdio.h>
#include <stdbool.h>
#include <gmp.h>
#include <string.h>
#include <stdlib.h>

#include "pe.h"

int sumDigitsFac(unsigned long n)
{
	mpz_t fn;
	char *digits;
	char digit[] = "x";
	int sum;

	mpz_init(fn);
	mpz_fac_ui(fn,n);

	digits = mpz_get_str(NULL,10,fn);
	sum = 0;
	for (int i = 0; i < strlen(digits); i++) {
		digit[0] = digits[i];
		sum += atoi(digit);
	}
	free(digits);
	mpz_clear(fn);
	return sum;
}

int main()
{
	pe_test_eq(sumDigitsFac(10), 27, "%d");

	printf("%d\n",sumDigitsFac(100));
}

File: go/eu020.go
// project euler (projecteuler.net) problem 20
// solution by Kevin Retzke (retzkek@gmail.com)
// December 2012
package main

import (
	"fmt"
	"math/big"
	"os"
)

func sumDigitsFac(num int64) int64 {
	n := new(big.Int)
	n.MulRange(1, num)

	ten := big.NewInt(10)
	d := new(big.Int)
	var sum int64

	for n.BitLen() > 0 {
		n.DivMod(n, ten, d)
		sum += d.Int64()
	}

	return sum
}

func main() {
	if r := sumDigitsFac(10); r != 27 {
		fmt.Printf("Test failed. sumDigitsFac(10) = %v (should be 27)\n", r)
		os.Exit(1)
	}
	fmt.Println(sumDigitsFac(100))
}
